modin/core/dataframe/pandas/exchange/dataframe_protocol/from_dataframe.py

```diff
@@ -0,0 +1,463 @@
+                columns[name], buf = convert_string_column(col)
+            elif dtype == DTypeKind.DATETIME:
+                columns[name], buf = convert_datetime_column(col)
+            else:
+                raise NotImplementedError(f"Data type {dtype} not handled yet")
+
+            buffers.append(buf)
+
+        pandas_df = pandas.DataFrame(columns)
+        pandas_df._buffers = buffers
+        return pandas_df
+
+    pandas_dfs = []
+    for chunk in df.get_chunks(nchunks):
+        pandas_df = get_pandas_df(chunk)
+        pandas_dfs.append(pandas_df)
+
+    pandas_df = pandas.concat(pandas_dfs, axis=0, ignore_index=True)
+
+    index_obj = df.metadata.get("modin.index", df.metadata.get("pandas.index", None))
+    if index_obj is not None:
+        pandas_df.index = index_obj
+
+    return pandas_df
+
+
+def convert_primitive_column_to_ndarray(col: ProtocolColumn) -> Tuple[np.ndarray, Any]:
+    """
+    Convert Column holding one of the primitive dtypes (int, uint, float or bool) to a NumPy array.
+
+    Parameters
+    ----------
+    col : ProtocolColumn
+
+    Returns
+    -------
+    tuple
+        Tuple of np.ndarray holding the data and the memory owner object that keeps the memory alive.
+    """
+    buffers = col.get_buffers()
+
+    data_buff, data_dtype = buffers["data"]
+    data = buffer_to_ndarray(data_buff, data_dtype, col.offset, col.size)
+
+    data = set_nulls(data, col, buffers["validity"])
+    return data, buffers
+
+
+def convert_categorical_column(col: ProtocolColumn) -> Tuple[pandas.Series, Any]:
+    """
+    Convert Column holding categorical data to a pandas Series.
+
+    Parameters
+    ----------
+    col : ProtocolColumn
+
+    Returns
+    -------
+    tuple
+        Tuple of pandas.Series holding the data and the memory owner object that keeps the memory alive.
+    """
+    ordered, is_dict, mapping = col.describe_categorical.values()
+
+    if not is_dict:
+        raise NotImplementedError("Non-dictionary categoricals not supported yet")
+
+    categories = np.array(list(mapping.values()))
+    buffers = col.get_buffers()
+
+    codes_buff, codes_dtype = buffers["data"]
+    codes = buffer_to_ndarray(codes_buff, codes_dtype, col.offset, col.size)
+
+    # Doing module in order to not get ``IndexError`` for out-of-bounds sentinel values in `codes`
+    values = categories[codes % len(categories)]
+
+    cat = pandas.Categorical(values, categories=categories, ordered=ordered)
+    data = pandas.Series(cat)
+
+    data = set_nulls(data, col, buffers["validity"])
+    return data, buffers
+
+
+def convert_string_column(col: ProtocolColumn) -> Tuple[np.ndarray, Any]:
+    """
+    Convert Column holding string data to a NumPy array.
+
+    Parameters
+    ----------
+    col : ProtocolColumn
+
+    Returns
+    -------
+    tuple
+        Tuple of np.ndarray holding the data and the memory owner object that keeps the memory alive.
+    """
+    if col.describe_null[0] not in (
```

Comment:
```suggestion
    null_kind, sentinel_val = col.describe_null
    if null_kind not in (
```
To not call multiple times describe_null, let's call it once and use the result further.